### DISCUSSION TITLE
blog/please-stop-using-barrel-files

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#107] [Announcements]
# blog/please-stop-using-barrel-files

Why you probably shouldn't be adding index.ts files everywhere

https://tkdodo.eu/blog/please-stop-using-barrel-files#inject-comments

**Comment by illright** [comment]
I fear that this article demonizes barrel files a bit too much. Issues like circular imports and pulling in more modules than you need are certainly worth considering, but they are not unique to barrel files. 

For example, if I have file that contains the config for some big DnD library and I also export some constants from it, then I get the same issue when I try to reach these constants.

I'm one of the maintainers of the [Feature-Sliced Design](https://feature-sliced.design) project, a frontend architecture methodology, and we rely quite heavily on barrel files to create a public API for a group of modules and protect the rest of the codebase from internal changes to that group of modules. The trick here is to not let that group of modules grow too big, or too diverse in its dependencies, then the dev performance downside of barrel files can be avoided. 

I personally think barrel files are a very useful concept with no reasonable alternative, and we should instead be pushing for bundlers to produce reasonable error messages (or just figure out circular imports wherever possible) and for better tooling that would help catch these issues early.

**Reply by ojj1123** [reply]
> be pushing for bundlers to produce reasonable error messages (or just figure out circular imports wherever possible) and for better tooling that would help catch these issues early.

Besides lint rule such as [import/no-cycle](https://github.com/import-js/eslint-plugin-import/blob/09476d7dac1ab36668283f9626f85e2223652b37/docs/rules/no-cycle.md), do you know there are any example tools that can help our figure out whether it affects development speed?

**Reply by TkDodo** [reply]
for nextJs, I just looked at the number of modules loaded during dev mode. Didn't find any tool that would tell what is loading which amount of modules though

**Reply by danielo515** [reply]
There is dependency-cruiser, which allows you to enforce import rules. Don't import from nested files, don't import from this module to this module, etc.

**Comment by madebyfabian** [comment]
What I am always asking myself is – in a Next.js app, I have a components/ui/index.ts barrel which exports about 40 components. Now all of these components are used in my application. I don't have any circular dependencies. Is it still beneficial for development performance to remove the barrel and import each UI component file separately instead? And are there any impacts on maybe script parsing time in production?

**Reply by TkDodo** [reply]
If this is a true barrel file that does nothing but re-export, `optimizePackageImports` in nextjs should be able to take care of this.

The circular imports can still happen if one of those components uses another, and the import is written to import from the barrel. The barrel looks convenient, but auto imports can pick up `import { Button } from 'components/ui/button' just as well as an import from the barrel. I don't see what the real advantage of the barrel is here ...

**Reply by landsman** [reply]
That's the issue with black magic in bundlers.

**Comment by fcasibu** [comment]
I avoid barrel files just because of potential side effects and use [imports field in package.json](https://nodejs.org/api/packages.html#imports) instead to make importing modules better

**Reply by fcasibu** [reply]
And [exports field](https://nodejs.org/api/packages.html#exports) if I'm exposing things publicly

**Comment by ryota-murakami** [comment]
Great post, I'd like to recommend [import/no-cycle](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-cycle.md) ESLint rule if you use barrel.

**Comment by PerryFinn** [comment]
🧐Without barrel files, there may be too many import statement issues, such as the following:
```js
// someComp.tsx
import A from "@/A";
import B from "@/A/B";
import C from "@/A/B/C";
//There are still many import statements ...

const Comp = ()=>{};
export {Comp};
```

**Reply by TkDodo** [reply]
what exactly are the issues with that?

- editors can collapse import sections so we don't see them
- usually, formatting will line-break after a couple of imports, so there is not much difference between:

```
import { Something } from '@/something'
import { SomethingElse } from '@/something-else'
import { SomethingElseEntirely } from '@/something-else-entirely'
```

and

```
import {
  Something, 
  SomethingElse,
  SomethingElseEntirely
} from '@'
```

it's actually fewer lines of code

**Reply by PerryFinn** [reply]
I completely agree with your second point that there is not much difference in the number of rows occupied after formatting.

Besides, I didn't actually know about the first point you mentioned. I just tried it out and sure enough, it can be folded in VSCode. Awesome!

**Comment by tranhl** [comment]
I largely agree with this position. However, would you still avoid barrel files when writing React folder components? Take this folder structure for example:

```
.
└── features/
    └── auth/
        ├── SignUpPage/
        │   ├── SignUpPage.tsx
        │   ├── ConfirmButton.tsx
        │   ├── Form.tsx
        │   └── index.ts 👈 Only exports `SignUpPage` from `./SignUpPage.tsx`
        ├── SignInPage/
        │   └── ...
        └── ...
```

The folder component pattern is nice here as it allows us to break out the `SignUpPage` component into smaller components while colocating all of the code for a feature. Without the barrel file, you'd have to import directly from `SignUpPage.tsx` like so:

```ts
import { SignUpPage } from '@/features/auth/SignUpPage/SignUpPage'
```

In this instance, having the barrel file provides two benefits:

1. It serves as the entry point for the folder component while also shortening the import path to `@/features/authentication/SignUpPage`
2. It allows us to codify what should/shouldn't be imported from this folder component by explicitly defining it in the barrel file.

Of course there's nothing stopping someone from directly importing `@/features/authentication/SignUpPage/ConfirmButton` for example. But as long as the team agrees to the convention, the tradeoffs are worthwhile.

This doesn't apply to just React folder components. I see this being useful when I want to break out large functions or classes into smaller modules, but still wanting to communicate the entry point for it all in code. 

Thoughts?

**Reply by TkDodo** [reply]
> Without the barrel file, you'd have to import directly from SignUpPage.tsx

Yes, but why is that bad? Even with the barrel file, consumers can still import the `SignUpPage` like this. They can also import the `ConfirmButton` directly, even though it's not part of the barrel. Feels like it's just a "convention" to only import from the index.ts file if it exists, but i's not something that is enforced.

I think the better convention is (and that is what we do), to set up a monorepo with different packages, then use `package.json` "exports" to define what the entry points are / what dependents can import. This would actually make things private / un-importable, and properly encapsulates the functionality. It has other advantages as well (when used with tools like nx or turborepo), like being able to skip a CI pipeline completely for a package if it hasn't been touched.

**Reply by tranhl** [reply]
I'm with you 100% on `package.json.exports` in a monorepo, that's definitely the way to go. But what if you're not as lucky, and work in a monolith? Then it's not quite as clear cut.

Let's say we want to add a new authentication page. If we avoid barrel files altogether and directly import from the source file, there's nothing stopping someone from deviating from the convention and creating `@/features/authentication/SignOutPage/Page.tsx` instead. Sure, we could conventionalize the file name of the entrypoint to match the folder, but then you lose out on shortening the import path.

Again, I totally agree that barrel files should be avoided if possible. I just wanted to highlight that they do have _some_ value when tackling specific problems.

**Reply by Sebastian-Nielsen** [reply]
@tranhl I would advice to use an eslint rule that enforces that `../` in an import path is not allowed to proceed an underscore. This way you can have some files in a folder be private to that folder. The con is of course an extra prepended character on a lot of files, but I'll take that any day over barrel files.

**Reply by danielo515** [reply]
You can use dependency cruise to enforce "only import from the barrel" on CI or in the pre-commit hook

**Reply by JamesPlayer** [reply]
> Yes, but why is that bad?

Aesthetics mainly, you get the folder/filename double ups on every import line (assuming each component has its own folder containing the component, a test file and hooks or sub-components):

```
import { SignUpPage } from '@/features/auth/SignUpPage/SignUpPage'
import { MyComponent } from '@/features/MyComponent/MyComponent'
import { MyOtherComponent } from '@/features/MyOtherComponent/MyOtherComponent'
```

**Reply by roberiof** [reply]
In this case you could add the code from SignUpPage.tsx inside index.tsx to avoid the import problem.

I thought that could cause problems for searching files or having multiple "index" files opened in the IDE at the same time, but that's not being a problem for me so far

**Comment by Thiagolino8** [comment]
Fortunately, in my experience, barrel files are a thing of the past; every decent bundler these days can handle them very well, at least for library code.
And by decent bundler, I mean none that next.js has ever used.

**Comment by lucaspere** [comment]
Your article was very convincing, but I was hoping for an alternative at the end to the barrel files. As you pointed out, they have improved a lot in organizing and maintaining the code. Do you have any alternative for this or just “stop using it”?

**Reply by TkDodo** [reply]
in a monorepo, you can organize into different packages, each having `exports` defined in `package.json` for everything that they expose publicly.

**Reply by jsPl** [reply]
But then how do you specify what should be exposed publicly if not by a barrel file?
I mean, something like
```json
  "exports": {
    "./feature": "./feature/index.js",
  }
```

**Reply by TkDodo** [reply]
you remove the `index` barrel and create multiple exports:

```
  "exports": {
    "./some-feature”: {
      "import": "./src/some-feature.js”
    },
    "./some-other-feature”: {
      "import": "./src/some-other-feature.js”
    },
  }
```

Then, you can import it with:

```
import {someFeature} from 'my-package/some-feature'
import {someOtherFeature} from 'my-package/some-other-feature'
```

**Reply by eglove** [reply]
You can also just use a directory as your "entry point" with tsup and others. Set it to src (or use the same as your includes array in your tsconfig). And your library will be structured as you structured it, the bundler will still chunk it and create imports as needed.

I think this is the right option for 99% of cases. I do it for all of my libraries.

If for whatever reason, you need to start hiding files (this shouldn’t be necessary or preferable in most cases), every good bundler like tsup still supports arrays.

I do not use package.json exports. No need.

**Reply by TkDodo** [reply]
Cool, I didn't know that

**Comment by kcrwfrd** [comment]
I think the example for your tab components are actually a good use-case for a barrel file, since they are likely always used together. The example of the circular import gotcha is just a caveat to be aware of: within a module (e.g. `@/tab`), don't import from that module's barrel file.

What you don't want to do is have a barrel file for an entire `@/components` dir.

**Reply by TkDodo** [reply]
in our design-system, we had several layers of barrel files:

- `index.ts`
  - re-exports * from `components.ts`, `utils.ts` etc
    - `components.ts` re-exports * from `tab.ts`, `button.ts` etc
      - `tab.ts` re-exports like 2 components and 1 hook

with that, it makes it very hard to see what the actual public interface of your component library is. It's nested and distributed around so many layers, it's not fun to navigate, you have to be super careful with what imports from where, auto-imports often get it wrong. The tab component could be imported in 4 ways:

```
import { Tab } from '@/design-system'
import { Tab } from '@/design-system/components'
import { Tab } from '@/design-system/components/tab'
import { Tab } from '@/design-system/components/tab/Tab'
```

And the first import pulls in 100+ components in dev mode. I found that avoiding / forbidding barrels altogether is the best approach :)

**Comment by mikitahimpel** [comment]
There is also a problem with barrel files when running tests. For example, we have a large nx application, nx provokes to create barrel files for modules, and since we import many other nx modules into the file that is covered by the unit test, jest has to compute this large file dependency graph for each unit test, and because of this we have most of the time spent on imports for unit tests.

**Reply by TkDodo** [reply]
Good point

**Comment by aaron-mota** [comment]
While I'm on team "avoid barrel files" because of their issues, one reason we decided to move to them previously (without all this knowledge of tradeoffs) is that, without barrel files, simply moving around a couple files/folders/components can turn what would be a small, couple-files-touched PR into tens or over a hundred files touched.

Maybe the problem is too young of a project, or just poor organization decisions, but what are your thoughts on that?

**Reply by TkDodo** [reply]
usually, IDEs can auto update imports in other files when moving a file around, but yeah it results in a larger PR.

I currently like monorepos with packages + "exports" defined in `package.json`. If you move a file, you can simply adapt the package.json and no consumer code must change.

**Comment by landsman** [comment]
It feels really good to write some stuff in Kotlin, you do not face premature issues like this. This problem should be fixed for years - in bundlers.  You should force bundlers and JS ecosystem to do the job better, not developers who just want to have clean imports. It's a shame that we read articles about same convetion issue over and over.

**Comment by blackhaj** [comment]
100% agree.

We have spent the last 6 months tidying up barrel files in our large monorepo (and still have plenty of work to do).

Do you have any tips for removing them en-masse? So far we are using some codemods to:
1. find out all the exports for a package and the sub-path of each
2. use the output of this in another codemod to update the places we import from that package

However, there are tons of edge cases and weird imports that get missed so it is very manual atm even with these codemods. My last PR which removed barrels for a mere two smaller, internal packages  was ~1.3k files 😅

I think part of the problem stems from the lack of info on how to properly structure a package to not allow imports. For a long time the "best practice" was to use the `main` property in your package.json but thankfully we now have `exports`

**Reply by TkDodo** [reply]
I've just used an eslint plugin and fixe the usages:

https://github.com/thepassle/eslint-plugin-barrel-files

**Comment by drvictor666** [comment]
Wow, wasn't aware of this issue and of the fact I'm a total barrel junkie :o However, now it's the time to change my ways. Very well written, thank you!

**Comment by kcrwfrd** [comment]
So, one question came up for me when investigating using `optimizePackageImports` in Next.js. The [documentation](https://nextjs.org/docs/app/api-reference/next-config-js/optimizePackageImports) seems to suggest that it's intended for use with external packages installed with NPM, but I'm wondering if it can also be used for barrel files defined within your Next.js app? How would you do that?

**Reply by TkDodo** [reply]
yes, I was doing that. Just point it towards your own packages, e.g. `@acme/my-package` and it will optimize it if it has an `index` entry point AND it is a real barrel file (that does nothing but export other things) that has no side effects or other `consts` declared

**Reply by kcrwfrd** [reply]
Hmm does it only work for packages with an entrypoint defined in a `package.json`? Or will it also work for random barrel files in our Next.js repo?

For example we have a `icons/index.ts` barrel file in our repo, and we have `{ "compilerOptions": { "baseUrl": ".", } }` in our `tsconfig.json` so that we can import like so

```ts
import { InstagramIcon, TikTokIcon } from 'icons'
```

So should this work with the following in our next config?

```ts
experimental: {
  optimizePackageImports: ['icons'],
},
```

My first attempt with this showed no change in bundle size from https://github.com/hashicorp/nextjs-bundle-analysis (but maybe I need to investigate more deeply?)

**Comment by eglove** [comment]
Are you sure libraries are somehow the exception? Because it's in apps where it tends to matter less. Lodash does internal packages and non-barrel imports just fine. It seems convenient to try to say @tanstack/react-query NEEDS a barrel file and no one else should be using them. And I'm confused as to why the main field of package.json is so important to you.

**Reply by TkDodo** [reply]
we could definitely remove `index.ts` and add each specific function as its own exports entry. Then users would import `useQuery` from a different location than `useMutation`. I would be fine with that, our users probably wouldn't 😂

**Reply by eglove** [reply]
How about both? @TkDodo I mean now that we've established that the only reason react-query does it is the same reason everyone else does it. For a minor convenience.

**Comment by varadan13** [comment]
we must be consious about what we are exporting out of a barel file. anything with unusual high import volume must be exported separately. i have seen people using barel files without knowing the performance cost of it

**Comment by wuarmin** [comment]
Hey @TkDodo,

thanks for this great post. 

I have a app with a folder-structure like this, and I would like to improve the file structure and naming convention. 
```txt
src/
├─ features/
│  ├─ orders/
│  │  ├─ order-addons/
│  │  │  ├─ order-addons.component.jsx
│  │  │  ├─ use-order-addons.js
```
So importing `OrderAddons`-component from `order-addons.component.jsx` is sub-optimal. 

```js
import { OrderAddons } from "@/features/order/order-addons/order-addons.component"
import { orderAddonsQueryOptions } from "@/features/order/order-addons/use-order-addons"
```
instead of 
```js
// using barrel files
import { OrderAddons, orderAddonsQueryOptions } from "@/features/order"
```
What would be your file structure convention (including naming) for this use-case?

Thanks

**Reply by TkDodo** [reply]
> So importing OrderAddons-component from order-addons.component.jsx is sub-optimal.

why? this is just a relative import:

```
// inside src/features/orders/order-addons/order-addons.component.jsx

import { orderAddonsQueryOptions } from './use-order-addons'
```

**Reply by wuarmin** [reply]
> why? this is just a relative import:

No, I posted the imports, I need outside. I.e. at the tanstack route, to preload the `orderAddons` with the help of `orderAddonsQueryOptions`

```js
import { OrderAddons } from "@/features/order/order-addons/order-addons.component"
import { orderAddonsQueryOptions } from "@/features/order/order-addons/use-order-addons"

export const Route = createFileRoute("/app/orders/$orderId/add-ons")({
  loader: ({ context: { queryClient }, params: { orderId } }) =>
    queryClient.ensureQueryData(orderAddonsQueryOptions({ orderId })),
  component: OrderAddons,
});
```

Which naming convention or file structure are you using? Are you using PascalCase names for components?

**Reply by wuarmin** [reply]
@TkDodo Could you please share your opinion on my question above?
I would be really interested. 
Thanks

**Comment by anicioalexandre** [comment]
Nice article! I've been trying to set the `exports` field on my packages but I can't get typescript auto imports to work. Did you guys manage to do that without barrel files? Is that even possible?

**Reply by TkDodo** [reply]
yeah this works in our monorepo. we have  `"type": "module"` and `exports` defined in `package.json`, and our base tsconfig defines `”pahts”` with `”moduleResolution": "Bundler"` and `“moduleDetection": "force"`.

**Comment by mimoid-prog** [comment]
So it looks like this is code editors issue that they can't handle barrels correctly. I know this is wishful thinking but if code editors auto-importing would prioritize to import from lowest possible barrel instead of direct file or some top level barrel then it all would be great.

**Reply by eglove** [reply]
Why is it the code editor? Are you talking about automatic imports? These aren't always available.

NextJS went through a lot of pain to resolve barrel files to be compatible with RSC during build. It's not even their responsibility.

I think it's library authors responsibility to stop using barrel files.

**Comment by RMHonor** [comment]
With ESM, you'll likely need the full path to the `index.js` anyway, so it doesn't even visually improve your imports much. Rather be explicit.

**Comment by chinanderm** [comment]
Why/how do the performance downsides not impact `@tanstack/react-query`? Why can libraries excluded from avoiding this practice?

**Reply by TkDodo** [reply]
They cannot, but it’s a trade-off that most consumers are’t willing to take. They mostly want a single file import structure. If you have icon libraries that export thousands of icons, that certainly is a problem, and frameworks like next go out of their way to automatically modularize those.

**Reply by eglove** [reply]
I really think people are more ok with long imports than you think. Big reason they're usually automatic. It's not often you have to manually write imports.

And people understand the benefits.

The number of people disagreeing that libraries are somehow the exception to barrel files here should be some sort of indicator.

**Comment by githorse** [comment]
I'm a little confused about the [Development Speed](https://tkdodo.eu/blog/please-stop-using-barrel-files#development-speed) point. In production, this would be mitigated by tree-shaking, such that only the exact files I need are actually downloaded (right?). So I guess this point is saying that in *development mode specifically*, my app will take a lot longer to start up if it has lots of barrel files? Or are we also saying there's potentially a hot module reload issue?

I may be totally misunderstanding this paragraph -- please somebody set me straight.

**Reply by TkDodo** [reply]
> In production, this would be mitigated by tree-shaking, such that only the exact files I need are actually downloaded (right?)

If your bundler supports treeshaking and it works correctly, yes. Could it be that unused things get shipped? Sure, maybe, but production bundle size wasn’t the point.

> So I guess this point is saying that in development mode specifically

Yes, the title is “Development Speed”, refering to how fast your app will start up in development mode.

> Or are we also saying there's potentially a hot module reload issue?

It will also negatively affect HMR, yes.

**Comment by harunskender97** [comment]
Thank you Dominik for this article.
Startup I am working on right now is super slow, it is nextJS 13.5 application and I've been investigating development issues.
We have 306 circular imports, HMR basically doesnt work how slow it is, and each time we change something in our codebase we need to wait for 30-60 seconds to see change and it reloads page. I will start with fixing this one

**Reply by TkDodo** [reply]
You have a good chance of fixing this by removing the barrels 👍

**Comment by Phenelo** [comment]
I ran into a tricky issue today with Next.js 15. It seems that Server Functions and Server Components can behave unexpectedly when they're in separate files and used together.

Specifically, if a Server Component calls a Server Function that relies on server-only code (like certain libraries or APIs), and that Server Component is re-exported through a barrel file that also re-exports Client Components (those with the "use client" directive), things can break.

I'm not entirely sure what's going on under the hood, but I suspect that including "use client" components in the same barrel file causes Next.js to treat the entire file (and maybe all exports) as client-side. This leads to errors when trying to use server-only code.

Removing the Server Component from the barrel file and importing it directly resolved the issue.

**Reply by TkDodo** [reply]
another reason not to use barrel files 😅

**Comment by Mostafa-dayasoft** [comment]
Besides the circular dependency issues caused by barrel files, I faced a more critical problem. We had an Element.ts barrel file that re-exported over 500 modules. In our main entry file (_app.tsx), we imported a single component through that barrel file , but the browser ended up downloading all 500 modules, totaling around 10MB.

After removing the barrel file and importing only what we needed directly, we saw a huge improvement in our page’s initial load performance.

**Reply by TkDodo** [reply]
seems like tree-shaking didn’t work out with your bundling setup ?

**Reply by ryota-murakami** [reply]
Circular dependency might be affected tree-shaking? 🤔

**Reply by TkDodo** [reply]
Could be.

**Reply by eglove** [reply]
The moral of the story is still don't use barrel files. The semantic of wanting cleaner imports is never worth it.

If you're using wilcard exports, files include side effects or globals, or exports aren't named explicitly, too much can be included in the bundle.

Instead of taking the time to try to carefully manage how, what, when, and where, just create direct named exports, no barrel files. If there are circular dependencies, a bundler will still be able to resolve and tree shake that.

Doesn't matter if it's an application or library there's no purpose for a barrel file.

**Reply by qwalker8408** [reply]
I think if you had to think of this in another way is to, with React, export from one module multiple components that can and should be composed together. So export component tab-list and tab-panel from tab. Much like shadcns approach to component design.

**Comment by bytesbybrandon** [comment]
I wasn't convinced. 

- Code review an catch enough of these issues. 
- AI code reviews can be told to look out for issues. 
- Good design helps with the development speed non-issue.

One thing that is consistent is that things will change. File locations, file names, organization of files, etc. The barrel file hides most of that within a "module" or however you're segregating and arranging your code base. Large PRs that are just a bunch of files being touched simply to update code paths because the designer got too creative is just not fun.

I agree with Monorepos doing a good job with limiting barrel file usage. That is just a nod back to good design.

**Reply by eglove** [reply]
Oh brother, we're officially in the age where the youngins are reading blogs from before AI existed and saying issues with existing lint rules can be caught by a probabilistic gold fish. 2026 everybody.

**Reply by bytesbybrandon** [reply]
If it makes you feel any more supercilious, I also mentioned code reviews by humans in addition to code reviews by AI.

2026 should be the year where we attack ideas rather than people. You didn't refute anything I said. You just passive aggressively attacked me. Not even youngins are doing that in 2026.

**Reply by eglove** [reply]
Nobody is attacking you. Learn static analysis before shoving everything into AI guidelines that may or may not be followed and depending on humans to review 10,000+ lines of duplicated code. It's not realistic.

By all research AI is a secondary tool to amplify existing tools. There is nothing you should be doing with AI that you don't and aren't already doing by hand and with existing tooling.

If you think your super detailed prompts and guidelines are the reason you're moving "10x" faster, I can promise you're not. You're moving 20% slower, producing at least 9% more bugs, increasing review time by 91%, and creating an absolute tech debt hell.

This blog is also not about AI. If you want to avoid barrel files, use ESLint. Do not tell AI to remind you not to use them. Do not depend on AI to run your pipelines for you. Use proper platform triggers. AI is a secondary addition. It doesn't replace anything at all. You still have to learn.

When you get called out on silly notions, trying to bring AI into something as simple as barrel files, and you get defensive, you're only cheating yourself.

**Comment by VesperQuartz** [comment]
Lol, was rethinking if I should rewrite my npm library and then I saw the last part 😅 but what about in monorepos?

**Reply by TkDodo** [reply]
I actually think that even libraries can avoid barrels by using multiple `exports` entries for things that they want to expose. The same is true for packages inside monorepos.

For example, [date-fns](https://github.com/date-fns/date-fns/blob/dd66398305c2b015fba3c1b3d31ccff42ee8d4cf/package.json#L63-L72) does it this way. Yes, they _also_ have a barrel file but I think we can probably move away from those for libs too.

This is another good read on this topic: https://dev.to/thepassle/a-practical-guide-against-barrel-files-for-library-authors-118c

**Reply by VesperQuartz** [reply]
i would take a look
